include/proxy-wasm/http_interface.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016-2019 Envoy Project Authors
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "include/proxy-wasm/compat.h"
+
+#include <time.h>
+#include <atomic>
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+#include "include/proxy-wasm/proxy_action.h"
+
+namespace proxy_wasm {
+
+/**
+ * HttpInterface is the interface between the VM host and the VM for HTTP streams.
+ */
+class HttpInterface {
+public:
+  /**
+   * Calls into the VM.
+   * These are implemented by the proxy-independent host code. They are virtual to support some
+   * types of testing.
+   */
+
+  /*
+   * HTTP
+   * Note: headers and trailers are obtained by the code in the VM via the XXXHeaderMapXXX set of
+   * functions. Body data is obtained by the code in the VM via calls implemented by the
+   * proxy-independent host code using the getBuffer() call.
+   */
+
+  /**
+   * Call on a stream context to indicate that the request headers have arrived.  Calls onCreate().
+   */
+  virtual ProxyAction onRequestHeaders() = 0;
```

Comment:
Why doesn't `onRequestHeaders()` also have an `end_of_stream` argument?

Reply:
Good point.  It doesn't need that because that information is cached in the Context object.  We don't need that information or the body_buffer_length either in onRequestBody() because the buffer is stored in the Context object as well so that it can handle calls out of the VM to retrieve bytes.   

IMO we should remove both of these from onRequestBody etc.

This would simplify the Context API.  The calls into the VM still contain that information, but they get the values form those cached in the Context.

Reply:
onRequestBody can be implemented from the getBuffer() and end_of_stream() calls on Context without those arguements in a generic way.

Reply:
Ack, it looks fine to remove all these args.